# Bonus:

In [2]:
# importing libraries

import numpy as np # fast numerical calculations
import pandas as pd # used for working with DataFrames (tables)
import matplotlib.pyplot as plt # creates visualization and charts
from sklearn.linear_model import LinearRegression # creates regression model

### Loading in CSV Files One-by-One

In [4]:
# file 1:
store_sales = pd.read_csv("Data/StoreSales.csv") # loading in StoreSales data

store_sales.info() # confirming all contents were imported

store_sales.head() #previewing table (top 5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 335129 entries, 0 to 335128
Data columns (total 5 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Transaction Date  335129 non-null  object 
 1   Store ID          335129 non-null  int64  
 2   RewardsID         34943 non-null   float64
 3   Prod Num          335129 non-null  object 
 4   Sale Amount       335129 non-null  float64
dtypes: float64(2), int64(1), object(2)
memory usage: 12.8+ MB


,Transaction Date,Store ID,RewardsID,Prod Num,Sale Amount
0,1/1/2022,702,NaN,105349-M,8.00
1,1/1/2022,704,NaN,105350-T,144.00
2,1/1/2022,705,NaN,105351-M,44.00
3,1/1/2022,705,NaN,105352-M,47.61
4,1/1/2022,705,NaN,105353-A,20.36


In [5]:
# file 2:
store_detail = pd.read_csv("Data/StoreDetail.csv") # loading in StoreDetail data

store_detail.info() # confirming all contents were imported

store_detail.head() # previewing table (top 5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 111 entries, 0 to 110
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Store Location     111 non-null    object
 1   State              111 non-null    object
 2   Store ID           111 non-null    int64 
 3   Territory Manager  111 non-null    object
 4   Region             111 non-null    object
 5   Region Director    111 non-null    object
dtypes: int64(1), object(5)
memory usage: 5.3+ KB


,Store Location,State,Store ID,Territory Manager,Region,Region Director
0,Aurora,Colorado,701,Jim Heck,West,Cassie Chambers
1,Berthoud,Colorado,702,Jim Heck,West,Cassie Chambers
2,Boulder,Colorado,703,Jim Heck,West,Cassie Chambers
3,Castle Rock,Colorado,704,Jim Heck,West,Cassie Chambers
4,Denver,Colorado,705,Jim Heck,West,Cassie Chambers


In [6]:
# file 3:
products = pd.read_csv("Data/Products.csv") # loading in Products data

products.info() # confirming all contents were imported

products.head() # previewing table (top 5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 669 entries, 0 to 668
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Prod Num       669 non-null    object
 1   Product        669 non-null    object
 2   CategoryID     669 non-null    int64 
 3   SubcategoryID  669 non-null    object
dtypes: int64(1), object(3)
memory usage: 21.0+ KB


,Prod Num,Product,CategoryID,SubcategoryID
0,105248-IT,TCL NXTPAPER 10s,120,120-tab
1,105249-IT,Dell Latitude 7320 Detachable,120,120-tab
2,105250-IT,Realme Pad,120,120-tab
3,105251-IT,Lenovo Tab P12 Pro,120,120-tab
4,105252-IT,Microsoft Surface Pro 9,120,120-tab


In [7]:
# file 4
product_categories = pd.read_csv("Data/ProductCategories.csv") # loading in ProductCategories data

product_categories.info() # confirming all contents were imported

product_categories.head() # previewing table (top 5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   CategoryID     52 non-null     int64 
 1   Category       52 non-null     object
 2   SubcategoryID  52 non-null     object
 3   Subcategory    52 non-null     object
dtypes: int64(1), object(3)
memory usage: 1.8+ KB


,CategoryID,Category,SubcategoryID,Subcategory
0,120,Technology & Accessories,120-tab,Tablets
1,120,Technology & Accessories,120-cal,Calculators
2,120,Technology & Accessories,120-sof,Software Download
3,120,Technology & Accessories,120-hea,Headphones
4,120,Technology & Accessories,120-ext,External Accessories


In [8]:
# file 5
customer_list = pd.read_csv("Data/customer_list.csv") # loading in customer_list data

customer_list.info() # confirming all contents were imported

customer_list.head() # previewing table (top 5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 521 entries, 0 to 520
Data columns (total 1 columns):
 #   Column                                           Non-Null Count  Dtype 
---  ------                                           --------------  ----- 
 0   cust_id|date|time|name|email|phone|sms-opt-out   521 non-null    object
dtypes: object(1)
memory usage: 4.2+ KB


,cust_id|date|time|name|email|phone|sms-opt-out
0,1|2023-03-15|08:45:12|Rachel|rachel@centralper...
1,2|2023-05-22|12:30:45|R. Geller|rossg@centralp...
2,3|2023-07-09|18:15:27|Monica Geller|chefmonica...
3,4|2023-09-01|21:05:33|Chandler Bing|chandlerb@...
4,5|2023-11-18|14:22:10|Joey|howyoudoing@central...


## Filtering, Sorting & Exploration

#### Q1:
Create a list of all transactions from January 2024, sorted by sale amount from highet to lowest. Which transactions had a sale amount greater than `$500`?

In [11]:
# converting Transaction Date to datetime
store_sales["Transaction Date"] = pd.to_datetime(store_sales["Transaction Date"]) # pulling month & year

# filtering for January 2024
jan_2024 = store_sales[(store_sales["Transaction Date"].dt.year == 2024) # checking for year 2024
                    & (store_sales["Transaction Date"].dt.month == 1)] # checking for month Jan

# sorting Janurary transactions from highest to lowest sale amount
jan_sort = jan_2024.sort_values(by="Sale Amount", ascending=False)

# displaying output for highest to lowest
jan_sort

,Transaction Date,Store ID,RewardsID,Prod Num,Sale Amount
135133,2024-01-09,872,NaN,105341-IT,1763.7
134987,2024-01-09,731,NaN,105341-IT,1763.7
139709,2024-01-31,822,NaN,105341-IT,1763.7
138780,2024-01-27,736,NaN,105341-IT,1763.7
138720,2024-01-27,701,NaN,105341-IT,1763.7
...,...,...,...,...,...
139643,2024-01-31,736,NaN,105444-S,2.4
133337,2024-01-01,901,324.0,105402-S,2.4
138904,2024-01-27,850,NaN,105611-S,2.4
136726,2024-01-17,739,NaN,105444-S,2.4


Time period starts from January 9, 2024 and ends at January 31, 2024.
They are sorted from highest total revenue at the top, and lowest at the bottom

In [13]:
# filtering Jan list to show transactions over $500 ONLY
jan_above_500 = jan_sort[jan_sort["Sale Amount"] > 500]

# displaying output for transactions over $500
jan_above_500

,Transaction Date,Store ID,RewardsID,Prod Num,Sale Amount
135133,2024-01-09,872,NaN,105341-IT,1763.7
134987,2024-01-09,731,NaN,105341-IT,1763.7
139709,2024-01-31,822,NaN,105341-IT,1763.7
138780,2024-01-27,736,NaN,105341-IT,1763.7
138720,2024-01-27,701,NaN,105341-IT,1763.7
...,...,...,...,...,...
139274,2024-01-29,815,NaN,105262-IT,603.5
139391,2024-01-30,718,NaN,105262-IT,603.5
136360,2024-01-15,847,NaN,105262-IT,603.5
134154,2024-01-05,825,NaN,105262-IT,603.5


The transactions that are over `$500` start at the highest with `$1,763.70` and the lowest with `$603.50`

#### Q2:
Find all products whose product number begins with `10525`. What category and subcategory do they belong to?

In [16]:
# creating new column to hold first 5 characters of product number
products["First_Five"] = products["Prod Num"].str[:5] # slicing the first 5 characters

# filtering products where first 5 characters equal 10525
prod_10525 = products[products["First_Five"] == "10525"] # matching

# merging product_categories to pull category & subcategory
prod_10525_info = prod_10525.merge(product_categories, # merging categories table
                                   on = ["CategoryID","SubcategoryID"]) # matching both IDs

# displaying output
prod_10525_info[["Prod Num","Product","Category","Subcategory"]]

,Prod Num,Product,Category,Subcategory
0,105250-IT,Realme Pad,Technology & Accessories,Tablets
1,105251-IT,Lenovo Tab P12 Pro,Technology & Accessories,Tablets
2,105252-IT,Microsoft Surface Pro 9,Technology & Accessories,Tablets
3,105253-IT,"Apple iPad Pro (12.9-inch, M2)",Technology & Accessories,Tablets
4,105254-IT,Chuwi HiPad Max,Technology & Accessories,Tablets
5,105255-IT,Lenovo Yoga Smart Tab,Technology & Accessories,Tablets
6,105256-IT,ASUS Vivobook 13 Slate OLED,Technology & Accessories,Tablets
7,105257-IT,Samsung Galaxy Tab S9 Ultra,Technology & Accessories,Tablets
8,105258-IT,Apple iPad Air (10.9-inch),Technology & Accessories,Tablets
9,105259-IT,Teclast T40 Pro,Technology & Accessories,Tablets


Products that begin with product number `10525` are in the Technology & Accessories Category, and in the Tablets Subcategory.

#### Q3:
Which stores in your assigned territories had at least one transaction with a sale amount over `$1,000`? How many such transactions did each store have?

In [19]:
# filtering store_detail only for SC & TX
assigned_stores = store_detail[(store_detail["State"] == "Texas")| # checking TX
                            (store_detail["State"] == "South Carolina")] # checking SC

# merging store_sales with assigned_stores on transaction
sales_w_state = store_sales.merge(assigned_stores, # merging filtered store list
                                  on = "Store ID") # matching Store ID

# filtering transactions with sale amount over $1,000
above_1000 = sales_w_state[sales_w_state["Sale Amount"] > 1000] # only above $1000

# counting how many are above 1000 transactions per store
store_count_1000 = above_1000.groupby(["Store ID","Store Location"] # grouping StoreID & Store Location
                                     )["Sale Amount"].count().reset_index() # counting amount of transactions

# renaming count column
store_count_1000 = store_count_1000.rename(columns={"Sale Amount": "Transactions Above $1,000"})

# displaying output
store_count_1000

,Store ID,Store Location,"Transactions Above $1,000"
0,852,Charleston,80
1,853,Greenville,98
2,901,Arlington,74
3,902,Austin,84
4,903,Bacliff,71
5,904,Baytown,73
6,905,Beaumont,183
7,906,Cedar Park,80
8,907,Dallas,11
9,908,Denton,82


#### Transactions Above $1,000
(SC) Charleston = 80, (SC) Greenville = 98
(TX) Arlington = 74, (TX) Austin = 84, (TX) Bacliff = 71, (TX) Baytown = 73, (TX) Beaumont = 183, (TX) Cedar Park = 80, (TX) Dallas = 11, (TX) Denton = 82, (TX) Desoto = 81, (TX) Fort Worth = 114, (TX) Georgetown = 69

## Aggregation & NumPy

#### Q4:
What is the total sales revenue across all in-store transactions in the dataset? What is the average transaction amount? What is the standard deviation?

In [23]:
# calculating total sales revenue per in-store transactions
total_revenue = store_sales["Sale Amount"].sum() # summing sale amounts

# calcualting average transaction amount per in-store transactions
avg_transaction = store_sales["Sale Amount"].mean() # avg sale amounts

# calculating standard deviation per transaction
std_transaction = store_sales["Sale Amount"].std() # standard deviation of sale amount

# displaying output
print(f"Total Revenue Per Store: {total_revenue:.2f}")
print(f"Average Transactions Per Store: {avg_transaction:.2f}")
print(f"Standard Deviation Per Sale: {std_transaction:.2f}")

Total Revenue Per Store: 45370048.85
Average Transactions Per Store: 135.38
Standard Deviation Per Sale: 279.72


#### Q5:
Which stores had total revenue above the average store revenue across the full dataset?

In [25]:
# calculating total revenue per store
store_totals = store_sales.groupby("Store ID")["Sale Amount"].sum().reset_index() # grouping stores and summing revenue

# calculating average store revenue across stores
avg_store_rev = np.mean(store_totals["Sale Amount"]) # grabbing average revenue

# filtering stores with above average revenue
above_avg_stores = store_totals[store_totals["Sale Amount"] > avg_store_rev] # keeping ONLY stores above avg

# merging store locations
above_avg_stores = above_avg_stores.merge(store_detail[["Store ID","Store Location"]], # adding city names per store
                                          on = "Store ID") # matching on Store ID

# displaying output
print(f"Average Total Revenue: {avg_store_rev:.2f}")
above_avg_stores

Average Total Revenue: 408739.18


,Store ID,Sale Amount,Store Location
0,705,917471.53,Denver
1,709,640368.11,Greeley
2,724,618846.25,Miami
3,728,541670.66,Tallahassee
4,734,584675.92,Germantown
5,736,8708119.00,North Harford
6,817,602183.44,Worcester
7,835,623163.53,Newark
8,838,597565.68,Trenton
9,847,607148.21,Rochester


#### Q6:
What is the total revenue for each product category across all territories, sorted from highest to lowest? Which category has the most individual transactions?